README.md
# README

## Run as a Developer

```sh
cd irc-link-history/public-html
docker run -it --name apache2 -p 80:80 -v $PWD/:/usr/local/apache2/htdocs/ httpd:2.4
cd ..
virualenv env   # Only needed the first time, or when new dependencies are added.
source env/bin/activate
pip install -r requirements.txt # Only needed when new dependencies are added.
cd app
uvicorn main:app --reload
# This is how you stop
docker stop apache2
docker container rm apache2
```

## Prepare for Production

```sh
cd irc-link-history/public-html
docker run -dit --name apache2 -p 80:80 -v $PWD/:/usr/local/apache2/htdocs/ httpd:2.4
# Make sure const uriToAPI is correct.
docker build --compress -t irc-link-history:0.1 . # Number corresponding to version number in Changelog
docker run --name irc-link-history -p 8000:8000 irc-link-history
# Make sure both localhost:8000 and localhost:8000 is working correctly.
docker save irc-link-history:latest -o irc-link-history.dockerfs
# scp over to the production server
```

## Run on Server

```sh
docker run --rm -dit --name apache2 -p 80:80 -v $PWD/:/usr/local/apache2/htdocs/ httpd:2.4
docker import irc-link-history -i irc-link-history.dockerfs
docker run --rm --name irc-link-history -p 8000:8000 irc-link-history
```

## Further Reading

MithrilJs tutorial [part 1](https://gilbert.ghost.io/mithril-js-tutorial-1/),  [part 2](https://gilbert.ghost.io/mithril-js-tutorial-2/) and [source code](https://github.com/gilbert/blog-post-examples/tree/gh-pages/mithril-2).


Changelog.md
# Changelog

## Version 0.121

* Fixed bug when initializing for the first time.

## Version 0.12 Working Prototype

* New posts on top
* Updated documentation

## Version 0.11 Working Prototype

* Can be manually deployed as a docker container
* Error handling of 404 errors when database is unreachable

## Version 0.1 Working Prototype

* Created backend using FastAPI (Python) and SQL
* Created frontend using MithrilJS, JavaScript and HTML
* Documentation consisting of roadmap, license, readme and changelog added
* Can only be run in a development environment


public-html/index.js
var root = document.body
var target = document.createElement("div")

const uriToAPI  = "http://ilmare.familjenberger.com:8000/" // For production
//const uriToAPI  = "http://localhost:8000/"  //For development

var SpotifyData = { list: [] }
var YoutubeData = { list: [] }
var PDFData     = { list: [] }
var AllitemsData ={ list: [] }
var APIData     = { list: [] }
var ErrorData   = { list: [] }

var Points     = {
    list: [ //"/" ,
            "/Spotify",
            "/Youtube",
            "/PDF",
            "/Allitems",
    ]
}


function getData(uri, struct) {
    m.request({
        url: uri ,
        method: "GET",
    })
    .then(function(result) {
        //console.log(result)
        /*
        console.log(JSON.parse(JSON.stringify(result.message[0])))
        console.log(result.message[0])
        result.message.forEach(element => {
            console.log(element)
        });
        */
        struct.list = result
    })
    .catch(function(e) {
        struct.error = e
        //console.log(struct.error.code)
        //console.log(struct.error.response.detail)
    })
}

function createParagraph(item) {
    var paragraph = document.createElement("P")
    paragraph.innerHTML = "<a target=\"_blank\" rel=\"noopener noreferrer\" href=\"" + item[1] + "\">" + item[0] + "</a>"
    document.getElementById("main").appendChild(paragraph)
}

function createErrorParagraph(code, details) {
    // console.log(item)
    var paragraph = document.createElement("P")
    paragraph.innerHTML = "The back end is probably on coffee break and static fall back pages are still not implemented. " + 
    "The Error code is " + code + "."
    //" and the message " + details
    document.getElementById("main").appendChild(paragraph)
}

function foo(){
}

var Start = {
    oninit: foo(),
    view: function(vnode) {
        document.getElementById("navbar").innerHTML = ""

        return Points.list.map(function(item) {
            //document.getElementById("main").innerHTML = "";
            //console.log(item)
            link = "#!/" + item
            title = item.split("/")
            title = title[1]
            var paragraph = document.createElement("div")
            paragraph.innerHTML = "<a href=\"" + link + "\">" + title + "</a>"
            document.getElementById("navbar").appendChild(paragraph)
    })
}
}

var Spotify = {
	oninit: getData(uriToAPI + "Spotify", SpotifyData),
    view: function(vnode) {
        if(SpotifyData.list.length == 0) {
            document.getElementById("main").innerHTML = ""
            createErrorParagraph(SpotifyData.error.code, SpotifyData.error.response.detail)
        }
        else {
            document.getElementById("main").innerHTML = ""
            return SpotifyData.list.map(function(item) {
                createParagraph(item)
            })
        }
    }
}

var Youtube = {
	oninit: getData(uriToAPI + "Youtube", YoutubeData),
    view: function(vnode) {
        if(YoutubeData.list.length == 0) {
            document.getElementById("main").innerHTML = ""
            createErrorParagraph(YoutubeData.error.code, YoutubeData.error.response.detail)
        }
        else {
            document.getElementById("main").innerHTML = ""
            return YoutubeData.list.map(function(item) {
                createParagraph(item)
            })
        }
    }
}

var PDFs = {
    oninit: getData(uriToAPI + "PDF", PDFData),
    view: function(vnode) {
        if(PDFData.list.length == 0) {
            document.getElementById("main").innerHTML = ""
            createErrorParagraph(PDFData.error.code, PDFData.error.response.detail)
        }
        else {
            document.getElementById("main").innerHTML = ""
            return PDFData.list.map(function(item) {
                createParagraph(item)
            })
        }
    }
}

var Allitems = {
    oninit: getData(uriToAPI + "Allitems", AllitemsData),
    view: function(vnode) {
        if(AllitemsData.list.length == 0) {
            document.getElementById("main").innerHTML = ""
            createErrorParagraph(AllitemsData.error.code, AllitemsData.error.response.detail)
        }
        else {
            document.getElementById("main").innerHTML = ""
            return AllitemsData.list.map(function(item) {
                createParagraph(item)
            })
        }
    }
}


//m.route(target, "/", {
m.route(target, "/Start", {
    "/Start": Start,
    "/Youtube": Youtube ,
    "/Spotify" : Spotify,
    "/PDF": PDFs,
    "/Allitems": Allitems
})
 
/*
    return m("a", {href: item[1]}, item[0])
    return m("a", {href: link}, title)
    return m("button", {onclick: function() {location.assign(link)}}, title ) //Fungerar
    return m.render(document.body, m("a", {href: item[1]}, item[0]))
    return m(m.route.Link, {href: item}, title)
    return m("div", item)
*/

/* Create a list.
    var node = document.createElement("LI")
    var textnode = document.createTextNode(item[0])
    node.appendChild(textnode)
    document.getElementById("links").appendChild(node)
*/



main.py
from sqlite3.dbapi2 import OperationalError
from typing import Optional

from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware

import sqlite3

from datetime import datetime
from pydantic.dataclasses import dataclass

@dataclass(frozen=True)
class linkCollection:
    title: str
    uri:   str
    time_created: datetime

app = FastAPI()

# https://fastapi.tiangolo.com/tutorial/cors/
origins = [
    "http://ilmare.familjenberger.com:8000",
    "https://ilmare.familjenberger.com:8000",
    "http://ilmare.familjenberger.com:80",
    "https://ilmare.familjenberger.com:80",
    "http://localhost:8000/",
    "https://localhost:8000/",
    "http://localhost:80/",
    "https://localhost:80/",
]

app.add_middleware(
    CORSMiddleware,
    #allow_origins=origins,
    allow_origins=['*'], 
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


try:
    connection = sqlite3.connect('history.dtek.db')
    # connection = sqlite3.connect(':memory:')
except FileNotFoundError as error:
    print(error)
    print("Database file not found.")
try: 
    cursor = connection.cursor()
    # cursor.execute("CREATE TABLE stuff(name text, num integer)")
    # cursor.execute("INSERT INTO stuff VALUES(?, ?)", ("Spotify", 3))
    # connection.commit()
except OperationalError as error:
    print(error)
    print("SQL operation failed.")


## @app.get("/")
## async def read_root():
##     return {"Message": "All your GET are belong to us!"}
##     #envarre = "%Spotify%"
##     #cursor.execute("SELECT * FROM stuff WHERE name LIKE ?", ("%Spotify%",))
##     #cursor.execute("SELECT * FROM stuff WHERE name LIKE ?", (envarre,))
## 
##     #data = cursor.fetchall()
##     #return(data)

@app.get("/Allitems")
async def readAllFromDB():
    cursor.execute('SELECT title, url FROM posts ORDER BY id DESC')
    data = cursor.fetchall()
    return(data)

@app.get("/PDF")
async def readPDFs():
    #cursor.execute('SELECT * FROM posts WHERE title LIKE '%YouTube%'')
    
    sql = """
    SELECT title, url FROM posts 
    WHERE url LIKE '%.pdf'
    ORDER BY id DESC"""
    cursor.execute(sql)

    data = cursor.fetchall()
    print(data)
    return(data)

@app.get("/Spotify")
async def readSpotify():

    # Original SQL search
    sql = """
    SELECT title, url FROM posts 
    WHERE title LIKE '%Spotify%'
    ORDER BY id DESC"""
    cursor.execute(sql)

    #envarre = "%Spotify%"
    #cursor.execute("SELECT title, url FROM posts WHERE title LIKE ?", (envarre,))

    data = cursor.fetchall()
    return(data)

@app.get("/Youtube")
async def readYouTube():
    #cursor.execute('SELECT * FROM posts WHERE title LIKE '%YouTube%'')
    
    sql = """
    SELECT title, url FROM posts 
    WHERE url LIKE '%youtube%' OR '%youtu.be%'
    ORDER BY id DESC"""
    cursor.execute(sql)

    data = cursor.fetchall()
    return(data)
    

# @app.get("/doc")
# async def read_root():
#     return {"Message": "All your GET are belong to us!"}
# 
# @app.get("/redoc")
# async def read_root():
#     return {"Message": "All your GET are belong to us!"}
